include/envoy/common/regex.h

```diff
@@ -0,0 +1,28 @@
+#pragma once
+
+#include <memory>
+
+#include "envoy/common/pure.h"
+
+#include "absl/strings/string_view.h"
+
+namespace Envoy {
+namespace Regex {
+
+/**
+ * A compiled regex expression matcher which uses an abstract regex engine.
+ */
+class CompiledMatcher {
```

Comment:
Why do we have both `StringMatcher` and `CompiledMatcher` with the same abstract interface? Would it makes sense to just have a `Matcher` that covers both?

Reply:
I thought about this, but I suspect that CompiledMatcher will eventually support other regex operations including extraction, etc. so I kept them separate. I don't feel strongly about this if you do though.

Reply:
Yeah, makes sense. Should they be part of a common hierarchy, since they both provide `match` though?

Reply:
Yes I can derive from the other class and add some comments.